### DISCUSSION TITLE
How to set provider with access token in the "providerOrUrl" of the HDWalletProvider instance.

### DISCUSSION THREAD

**Discussion by ksanjaykumar1** [ConsenSys-archive/truffle#5798] [General Q&A] [answered]
I have install security-plugin in the goQuorum private network, I have to send access token while communicating with the node. 

I am not able to migrate contracts. I not able to set the correct provider thus failing to migrate. I confirmed network is working correctly by getting block number using web3js library. 

here is my truffle-config.js 

    const HDWalletProvider = require('@truffle/hdwallet-provider');
    
    const fs = require('fs');
    const privateKey = fs.readFileSync('.private-key').toString().trim();
    
    // creating provider 
    require('dotenv').config();
    let Web3 = require('web3');
    const { TOKEN, BLOCKCHAIN_NODE } = process.env;
    let HttpHeaderProvider = require('httpheaderprovider');
    let headers = { Authorization: `Bearer ${TOKEN}` };
    console.log(headers)
    
    module.exports = {
     
    
      contracts_directory: './contracts/permissions',
    
      networks: {
        development: {
          provider: function () {
            return new HDWalletProvider({
              privateKeys: [privateKey],
              providerOrUrl: new Web3.providers.HttpProvider(BLOCKCHAIN_NODE,headers),
              chainId: 1337,
            });
          }, // replace with quorum node port you wish to connect to
          network_id: '*',
          type: 'quorum',
          gasPrice: 0,
          gas: 450000000,
        },
      },
    
      // Set default mocha options here, use special reporters etc.
      mocha: {
        // timeout: 100000
      },
    
      // Configure your compilers
      compilers: {
        solc: {
          version: '0.5.3', // Fetch exact version from solc-bin (default: truffle's version)
          // docker: true,        // Use "0.5.1" you've installed locally with docker (default: false)
          // settings: {          // See the solidity docs for advice about optimization and evmVersion
          //  optimizer: {
          //    enabled: false,
          //    runs: 200
          //  },
          //  evmVersion: "byzantium"
          // }
        },
      },
    };

I have also tried by creating provider with httpheaderprovider and passing it 
        
    const HDWalletProvider = require('@truffle/hdwallet-provider');
    
    const fs = require('fs');
    const privateKey = fs.readFileSync('.private-key').toString().trim();
    
    // creating provider 
    require('dotenv').config();
    const { TOKEN, BLOCKCHAIN_NODE } = process.env;
    let HttpHeaderProvider = require('httpheaderprovider');
    let headers = { Authorization: `Bearer ${TOKEN}` };
    const provider = new HttpHeaderProvider(BLOCKCHAIN_NODE, headers);
    
    
    module.exports = {
      /**
       * Networks define how you connect to your ethereum client and let you set the
       * defaults web3 uses to send transactions. If you don't specify one truffle
       * will spin up a development blockchain for you on port 9545 when you
       * run `develop` or `test`. You can ask a truffle command to use a specific
       * network from the command line, e.g
       *
       * $ truffle test --network <network-name>
       */
    
      contracts_directory: './contracts/permissions',
    
      networks: {
        development: {
          provider: function () {
            return new HDWalletProvider({
              privateKeys: [privateKey],
              providerOrUrl: provider,
              chainId: 1337,
            });
          }, // replace with quorum node port you wish to connect to
          network_id: '*',
          type: 'quorum',
          gasPrice: 0,
          gas: 450000000,
          
        },
      },
    
      // Set default mocha options here, use special reporters etc.
      mocha: {
        // timeout: 100000
      },
    
      // Configure your compilers
      compilers: {
        solc: {
          version: '0.5.3', // Fetch exact version from solc-bin (default: truffle's version)
          // docker: true,        // Use "0.5.1" you've installed locally with docker (default: false)
          // settings: {          // See the solidity docs for advice about optimization and evmVersion
          //  optimizer: {
          //    enabled: false,
          //    runs: 200
          //  },
          //  evmVersion: "byzantium"
          // }
        },
      },
    

    };

I am attaching the screenshots which I got. 
1. By creating provider with Web3 
![Screenshot from 2022-12-17 12-07-26](https://user-images.githubusercontent.com/72605368/208230502-dd72b810-a30a-4ab0-becb-c4607d5d1e4b.png)
2. By creating provider with httpheaderprovider
![Screenshot from 2022-12-16 14-46-14](https://user-images.githubusercontent.com/72605368/208230556-93283c5a-5dcb-486d-b033-183f65461d14.png)

3. By passing just the url ( I got two different types of errors when i ran  multiple times)
![Screenshot from 2022-12-16 14-46-53](https://user-images.githubusercontent.com/72605368/208230603-f419ce02-6a88-402e-8bbd-10b8aa8b67ea.png)
![Screenshot from 2022-12-16 14-53-39](https://user-images.githubusercontent.com/72605368/208230610-d6b5ad83-8249-4557-a4ed-335e68e5f236.png)

5. Confirmation that plugin is working and able to get blocknumber 
![Screenshot from 2022-12-16 14-47-47](https://user-images.githubusercontent.com/72605368/208230644-b7b06fcb-3dac-4155-8db6-704ca7869e45.png)

**Comment by cds-amal** [comment, marked as answer]
Hi, [httpheaderprovider](https://www.npmjs.com/package/httpheaderprovider) is pretty old, and doesn't seem to have much support or usage. Have you tried passing an options object that Web3 expects? [See their header tests](https://github.com/web3/web3.js/blob/316bb6ca0d8aaea56c824efb517c2db1286bba46/test/httpprovider.js#L36), and [docs](https://web3js.readthedocs.io/en/v1.8.1/include_package-core.html?highlight=Web3HttpProvider#configuration)

**Reply by ksanjaykumar1** [reply]
I think it worked correctly. But I am getting a new error which says "Error: net_version - access denied". I checked the version by using geth command line in the node and it says "1337" and I passed ,  chainId: 1337, in HDWalletProvider instance. 

![Screenshot from 2022-12-19 11-21-58](https://user-images.githubusercontent.com/72605368/208357216-4c923862-82e4-475d-a0ea-85a216d0317b.png)
![Screenshot from 2022-12-19 11-06-52](https://user-images.githubusercontent.com/72605368/208357342-d4f8b892-14b2-425b-805a-e36e31bdcaab.png)


complete error message is 

    Error: net_version - access denied
        at /home/sanjay/csg/vcn-development/quire-contracts-development/node_modules/@trufflesuite/web3-provider-engine/subproviders/provider.js:18:36
        at /home/sanjay/csg/vcn-development/quire-contracts-development/node_modules/web3-providers-http/lib/index.js:124:13
        at processTicksAndRejections (node:internal/process/task_queues:96:5)
    Truffle v5.0.42 (core: 5.0.42)
    Node v16.18.1

My truffle-config.js is 

    const HDWalletProvider = require('@truffle/hdwallet-provider');
    
    const fs = require('fs');
    const privateKey = fs.readFileSync('.private-key').toString().trim();
    
    // creating provider
    require('dotenv').config();
    const { TOKEN, BLOCKCHAIN_NODE } = process.env;
    var Web3HttpProvider = require('web3-providers-http');
    var options = {
      headers: [
        { name: 'Access-Control-Allow-Origin', value: '*' },
        { name: 'Authorization', value: `Bearer ${TOKEN}` },
      ],
    };
    var provider = new Web3HttpProvider(BLOCKCHAIN_NODE, options);
    
    module.exports = {
    
      contracts_directory: './contracts/permissions',
    
      networks: {
        development: {
          provider: function () {
            return new HDWalletProvider({
              privateKeys: [privateKey],
              providerOrUrl: provider,
              chainId: 1337,
            });
          }, // replace with quorum node port you wish to connect to
          network_id: '*',
          type: 'quorum',
          gasPrice: 0,
          gas: 450000000,
        },
      },
    
      // Set default mocha options here, use special reporters etc.
      mocha: {
        // timeout: 100000
      },
    
      // Configure your compilers
      compilers: {
        solc: {
          version: '0.5.3', // Fetch exact version from solc-bin (default: truffle's version)
          // docker: true,        // Use "0.5.1" you've installed locally with docker (default: false)
          // settings: {          // See the solidity docs for advice about optimization and evmVersion
          //  optimizer: {
          //    enabled: false,
          //    runs: 200
          //  },
          //  evmVersion: "byzantium"
          // }
        },
      },
    };

**Reply by ksanjaykumar1** [reply]
Thank You. Now I able to migrate contracts. "net_version - access denied" error was due to "rpc://*_version" not being included in the scope of access token.
